### ISSUE TITLE
ArduinoIoTCloudClass::update() stuck when internet connection fails

### ISSUE THREAD

**Issue by raffaelecassia** [arduino-libraries/ArduinoIoTCloud#86]
Hi all,
I'm having trouble keeping a nano33iot running when the internet connection (but *not* wifi) fails/drops for a short period of time.
It gets stuck in the function `void ArduinoIoTCloudClass::update(CallbackFunc onSyncCompleteCallback)` called from the loop() code.

After the connection fails, the board will have 45/60 seconds of life before getting completely stuck inside that function. No trying to reconnect occurs.
Even if the internet connection fails for only a couple of seconds, the board will freeze in 45/60 seconds.
The only solution is to reset either the board or the access point (the latter will trigger a wifi drop and the board will come back to life again searching for the wifi network). 

Tested repeatedly in the last few days in many ways:
- unplugging the ethernet cable on a wifi accesspoint
- disabling the mobile network on android in hotspot mode
- casually walking away from the AP with the board in hands (this sometimes will trigger the wifi drop+reconnection)
- different accesspoint/networks
- on a mkrwifi1010

With some Serial.print-fu I was able to narrow it a little bit to the call `_mqttClient->poll()` inside the update() function. 

Tested sketch is fresh template generated by the Arduino Create IoT Cloud compiled with the latest release of ArduinoIoTCloud (and dependencies).

**Comment by raffaelecassia** [comment]
Ok after some digging I found something more.

Here's a (sort of) stacktrace when the board freeze:

- `br_sslio_close(br_sslio_context)`   https://github.com/arduino-libraries/ArduinoBearSSL/blob/master/src/bearssl/ssl_io.c#L305
- `BearSSLClient::stop()`   https://github.com/arduino-libraries/ArduinoBearSSL/blob/master/src/BearSSLClient.cpp#L154
- `MqttClient::stop()`   https://github.com/arduino-libraries/ArduinoMqttClient/blob/master/src/MqttClient.cpp#L740
- `MqttClient::poll()`   https://github.com/arduino-libraries/ArduinoMqttClient/blob/master/src/MqttClient.cpp#L620
- `ArduinoIoTCloudClass::update()`   https://github.com/arduino-libraries/ArduinoIoTCloud/blob/master/src/ArduinoIoTCloud.cpp#L232

If I comment out the `while` ("Discard any incoming application data.") from `br_sslio_close()` function, everything will work as expected.
I know, this isn't an elegant solution. At least I don't see memory leaks in my tests.

**Comment by aentinger** [comment]
@sandeepmistry Do you have any idea about this issue? Maybe we should add a timeout at https://github.com/arduino-libraries/ArduinoBearSSL/blob/master/src/bearssl/ssl_io.c#L305 ?

**Comment by kaiomatico** [comment]
I am experiencing the same issue. Is there any workaround to this problem like
1.: Surrounding the update function with a try-catch code block
2.: A timer interrupt as second thread which restarts the board every 30 minutes or so as precaution? 
Thank you all very much in advance  :+1:

**Comment by sircambridge** [comment]
I would like to know if there is a workaround as well. Even if I can configure some sort of timeout threshold would be ok - currently my device will hang for up to a minute per loop when the wifi is out of range, which makes everything very awkward.

**Comment by adamansky** [comment]
Hi! All I've found a workaround:

0. Declare `void br_ssl_engine_fail(br_ssl_engine_context *cc, int err);` in your module.
1. Set some flag boolean flag (let it be `sslio_closing`) indicating thet you are going to close bearssl context (`br_sslio_close(br_sslio_context)`)
2. In read and write functions attached to context do the following checks like this:
```c
for (;;) {
    ssize_t wlen;
    wlen = write(fd, buf, len);
    if (wlen <= 0) {
      if (sslio_closing) { // <---- Check you are closing context
        br_ssl_engine_fail(session->ioc.engine, BR_ERR_IO);
      } else if (wlen < 0 && (errno == EINTR || errno == EAGAIN)) {
        continue;
      }
      return -1;
    }
    return (int)wlen;
  }
```
It helps me to avoid suchlike hangs.

I've tried to contact Thomas Porning concerning this issue but had no success.

Good luck!

**Comment by IgorSchelle** [comment]
Same Issue for my Arduino wifi mkr 1010.  As it usually gets stuck within a few minutes, it makes my two boards completely useless.
Is there any plan for fixing this? I can help out with some testing.

**Comment by simonegenovese** [comment]
Same here! Please fix!

**Comment by aentinger** [comment]
Hi :wave: We've got new [releases](https://github.com/arduino/nina-fw/releases) of the firmware running on the WiFi module (NINA-W102). Would you mind trying to upgrade to the latest release ([1.4.3](https://github.com/arduino/nina-fw/releases/tag/1.4.3)) and see if that solves your issue? Best way to do is is probably: https://github.com/arduino/FirmwareUpdater.

**Comment by IgorSchelle** [comment]
Hi, I have upgraded FW to 1.4.3 and WiFiNINA - Version: 1.8.2. I have uploaded a simple sketch that just controlled inbuild led from my dashboard.  It runs just fine for days (at least 14).  It was promising so  I took another MKR 1010 with FW 1.4.3 and library 1.8.3 upload sketch where Arduino reports each second temperature and humidity from a sensor. This died after few minutes. I mean, the main loop just stopped. And the same happened later also with the first simple code.  So the problem is still there it just happens later.  One interesting thing i noticed. My first arduino appeared to be offline in device list but I was still able to controll it.
